File: smells/unreliable/fire-and-forget.js
/* Запах: Fire and Forget
 */

// Тестируемый модуль
function loadUser (id, cb) {
  var path = '/users/' + id

  get(path, function (er, user) {
    user.resolvedVia = path
    cb(er, user)
  })
}

// Тесты
module.exports = {
  getsUserAndDecoratesPath: function () {
    loadUser(42, function (er, user) {
      assert.equal(user.resolvedVia, '/users/42')
      assert.equal(user.name, 'Jo')
    })
  }
}

// Фейковая реализация
function get (path, cb) {
  setTimeout(function () {
    cb(null, {name: 'Jo'})
  }, 10)
}

File: smells/unreliable/fire-and-forget/fire-and-forget.rb
# Subject under test
def load_user(id)
  path = "/users/#{id}"

  get(path) { |er, user|
    user.resolved_via = path
    yield er, user if block_given?
  }
end

# Test
class FireAndForget < SmellTest
  def test_gets_user_and_decorates_path
    load_user(42) { |er, user|
      assert_equal "/users/42", user.resolved_via
      assert_equal "Jo", user.name
    }
  end
end

# Fake production implementations to simplify example test of subject
def get(path)
  Thread.new do
    sleep 0.01 if rand < 0.5 # sometimes it takes time
    yield nil, OpenStruct.new(name: "Jo") if block_given?
  end
end

